libopae++/src/handle.cpp

```diff
@@ -33,41 +34,88 @@ namespace fpga
 namespace types
 {
 
-handle::handle(fpga_handle h)
+handle::handle(fpga_handle h, uint32_t mmio_region, uint8_t *mmio_base)
 : handle_(h)
+, mmio_region_(mmio_region)
+, mmio_base_(mmio_base)
 {
 }
 
 handle::~handle(){
     close();
 }
 
+bool handle::write(uint64_t offset, uint32_t value)
+{
+    return FPGA_OK == fpgaWriteMMIO32(handle_, mmio_region_, offset, value);
+}
+
+bool handle::write(uint64_t offset, uint64_t value)
+{
+    return FPGA_OK == fpgaWriteMMIO64(handle_, mmio_region_, offset, value);
+}
+
+bool handle::read(uint64_t offset, uint32_t &value) const
+{
+    return FPGA_OK == fpgaReadMMIO32(handle_, mmio_region_, offset, &value);
+}
+
+bool handle::read(uint64_t offset, uint64_t &value) const
+{
+    return FPGA_OK == fpgaReadMMIO64(handle_, mmio_region_, offset, &value);
+}
+
+handle::ptr_t handle::open(fpga_token token, int flags, uint32_t mmio_region){
+    fpga_handle c_handle = nullptr;
+    uint8_t *mmio_base = nullptr;
+    ptr_t p;
 
-handle::ptr_t handle::open(fpga_token token, int flags){
-    fpga_handle c_handle;
     auto res = fpgaOpen(token, &c_handle, flags);
-    if (res == FPGA_OK){
-        return handle::ptr_t(new handle(c_handle));
+    if (res == FPGA_OK) {
+
+        res = fpgaMapMMIO(c_handle, mmio_region, reinterpret_cast<uint64_t **>(&mmio_base));
+
+        if (res == FPGA_OK) {
+           p.reset(new handle(c_handle, mmio_region, mmio_base));
+        } else {
+           res = fpgaClose(c_handle);
+           // TODO : Log or throw error
+        }
     }
-    // TODO : Log or throw error
-    return handle::ptr_t();
+
+    return p;
 }
 
-handle::ptr_t handle::open(token::ptr_t token, int flags){
-    return handle::open(token->get(), flags);
+handle::ptr_t handle::open(token::ptr_t token, int flags, uint32_t mmio_region){
+    return handle::open(token->get(), flags, mmio_region);
 }
 
 fpga_result handle::close(){
+    fpga_result res = FPGA_INVALID_PARAM;
+
     if (handle_ != nullptr){
-        auto res = fpgaClose(handle_);
-        if (res == FPGA_OK){
-            handle_ = nullptr;
-        }else{
+
+        res = fpgaUnmapMMIO(handle_, mmio_region_);
```

Comment:
Should we only unmap if the base ptr is non null?

Reply:
I don't think it's necessary. The handle tracks which regions have been allocated, and the call will fail if region mmio_region_ is not found.